IllegalArgumentException using bmc ObjectStorage client with micronaut 4, when there are special chars in object names

### Expected Behavior

I should be able to programmatically create Objects, with arbitrary names including special characters like "|"

### Actual Behaviour

Produces this error:

>
> java.lang.IllegalArgumentException: Illegal character in path at index 89: https://objectstorage.us-phoenix-1.oraclecloud.com/n/mynamespace/b/mybucket/o/path%2Fwith|pipe
> 	at java.base/java.net.URI.create(URI.java:932)
> 	at io.micronaut.oraclecloud.httpclient.netty.NettyHttpRequest.uri(NettyHttpRequest.java:205)
> 	at com.oracle.bmc.http.internal.ClientCall.buildServiceDetails(ClientCall.java:173)
> 	at com.oracle.bmc.http.internal.ClientCall.callSync(ClientCall.java:1129)
> 	at com.oracle.bmc.objectstorage.ObjectStorageClient.getObject(ObjectStorageClient.java:1018)
> 	at ddifranc.ObjectStoreNameTest.testGet(ObjectStoreNameTest.java:25)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
> 	at io.micronaut.test.extensions.junit5.MicronautJunit5Extension$2.proceed(MicronautJunit5Extension.java:142)
> 	at io.micronaut.test.extensions.AbstractMicronautExtension.interceptEach(AbstractMicronautExtension.java:162)
> 	at io.micronaut.test.extensions.AbstractMicronautExtension.interceptTest(AbstractMicronautExtension.java:119)
> 	at io.micronaut.test.extensions.junit5.MicronautJunit5Extension.interceptTestMethod(MicronautJunit5Extension.java:129)
> 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
> 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
> Caused by: java.net.URISyntaxException: Illegal character in path at index 89: https://objectstorage.us-phoenix-1.oraclecloud.com/n/mynamespace/b/mybucket/o/path%2Fwith|pipe
> 	at java.base/java.net.URI$Parser.fail(URI.java:2995)
> 	at java.base/java.net.URI$Parser.checkChars(URI.java:3166)
> 	at java.base/java.net.URI$Parser.parseHierarchical(URI.java:3248)
> 	at java.base/java.net.URI$Parser.parse(URI.java:3196)
> 	at java.base/java.net.URI.<init>(URI.java:645)
> 	at java.base/java.net.URI.create(URI.java:930)
> 	... 12 more

### Steps To Reproduce

```
@Test
void testGet() {
GetObjectRequest getRequest = GetObjectRequest.builder()
       .namespaceName(“mynamespace”)
       .bucketName(“mybucket”)
       .objectName(“path/with|pipe”).build();
GetObjectResponse response = osclient.getObject(getRequest);
```

//see example application for full test

### Environment Information

MacOS
Java 17

**Non-working configuration:**
io.micronaut:micronaut-core:jar:4.6.6
io.micronaut.oraclecloud:micronaut-oraclecloud-bmc-objectstorage:jar:4.2.1
com.oracle.oci.sdk:oci-java-sdk-objectstorage:jar:3.47.0

**working configuration (in which OCI SDK uses JerseyHttpRequest rather than netty):**
io.micronaut:micronaut-core:jar:3.9.4
com.oracle.oci.sdk:oci-java-sdk-objectstorage:jar:2.71.1



### Example Application

https://github.com/ddifranc/micronaut-objectstore-illegal-names

### Version

4.6.6

--- a/oraclecloud-atp-hikari-test/src/test/groovy/io/micronaut/oraclecloud/atp/jdbc/hikari/HikariPoolConfigurationListenerSpec.groovy
+++ b/oraclecloud-atp-hikari-test/src/test/groovy/io/micronaut/oraclecloud/atp/jdbc/hikari/HikariPoolConfigurationListenerSpec.groovy
@@ -40,6 +40,9 @@ class HikariPoolConfigurationListenerSpec extends Specification {
         ResultSet resultSet = connection.createStatement().executeQuery("SELECT * FROM DUAL")
         resultSet.next()
         resultSet.getString(1) == "X"
+
+        cleanup:
+        context.close()
     }
 
     def "test it skips datasource without ocid field"() {
@@ -58,5 +61,8 @@ class HikariPoolConfigurationListenerSpec extends Specification {
         ResultSet resultSet = connection.createStatement().executeQuery("SELECT 1")
         resultSet.next()
         resultSet.getString(1) == "1"
+
+        cleanup:
+        context.close()
     }
 }

--- a/oraclecloud-common/src/test/groovy/io/micronaut/discovery/cloud/OracleCloudV2MetadataCertSpec.groovy
+++ b/oraclecloud-common/src/test/groovy/io/micronaut/discovery/cloud/OracleCloudV2MetadataCertSpec.groovy
@@ -50,8 +50,8 @@ class OracleCloudV2MetadataCertSpec extends Specification {
 
         cleanup:
         ctx.close()
-        server.stop()
-        serverNew.stop()
+        server.close()
+        serverNew.close()
     }
 
     @Controller("/opc/v2")

--- a/oraclecloud-common/src/test/groovy/io/micronaut/discovery/cloud/OracleCloudV2MetadataResolverSpec.groovy
+++ b/oraclecloud-common/src/test/groovy/io/micronaut/discovery/cloud/OracleCloudV2MetadataResolverSpec.groovy
@@ -47,7 +47,7 @@ class OracleCloudV2MetadataResolverSpec extends Specification {
 
         cleanup:
         ctx.close()
-        server.stop()
+        server.close()
     }
 
     @Controller("/opc/v2")

--- a/oraclecloud-httpclient-apache-http-core/src/main/java/io/micronaut/oraclecloud/httpclient/apache/core/ApacheCoreHttpRequest.java
+++ b/oraclecloud-httpclient-apache-http-core/src/main/java/io/micronaut/oraclecloud/httpclient/apache/core/ApacheCoreHttpRequest.java
@@ -45,7 +45,6 @@ import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
 import java.net.URI;
-import java.net.URISyntaxException;
 import java.net.UnixDomainSocketAddress;
 import java.nio.channels.Channels;
 import java.nio.channels.SocketChannel;
@@ -138,6 +137,9 @@ final class ApacheCoreHttpRequest implements HttpRequest {
 
     @Override
     public HttpRequest appendPathPart(String encodedPathPart) {
+        // java-sdk follows whatwg spec for escaping, but RFC 2396 (followed by java.net.URI) does not permit pipe characters
+        encodedPathPart = encodedPathPart.replace("|", "%7c");
+
         boolean hasSlashLeft = !path.isEmpty() && path.charAt(path.length() - 1) == '/';
         boolean hasSlashRight = encodedPathPart.startsWith("/");
         if (hasSlashLeft) {
@@ -177,11 +179,7 @@ final class ApacheCoreHttpRequest implements HttpRequest {
 
     @Override
     public URI uri() {
-        try {
-            return new URI(client.baseUri.getScheme(), client.baseUri.getAuthority(), path.toString(), query.isEmpty() ? null : query.toString(), null);
-        } catch (URISyntaxException e) {
-            throw new IllegalStateException(e);
-        }
+        return URI.create(client.baseUri + buildRelativeUri());
     }
 
     @Override

--- a/oraclecloud-httpclient-netty/src/main/java/io/micronaut/oraclecloud/httpclient/netty/MicronautHttpRequest.java
+++ b/oraclecloud-httpclient-netty/src/main/java/io/micronaut/oraclecloud/httpclient/netty/MicronautHttpRequest.java
@@ -177,6 +177,9 @@ final class MicronautHttpRequest implements HttpRequest {
 
     @Override
     public HttpRequest appendPathPart(String encodedPathPart) {
+        // java-sdk follows whatwg spec for escaping, but RFC 2396 (followed by java.net.URI) does not permit pipe characters
+        encodedPathPart = encodedPathPart.replace("|", "%7c");
+
         boolean hasSlashLeft = uri.charAt(uri.length() - 1) == '/';
         boolean hasSlashRight = encodedPathPart.startsWith("/");
         if (hasSlashLeft) {

--- a/oraclecloud-httpclient-netty/src/main/java/io/micronaut/oraclecloud/httpclient/netty/NettyHttpRequest.java
+++ b/oraclecloud-httpclient-netty/src/main/java/io/micronaut/oraclecloud/httpclient/netty/NettyHttpRequest.java
@@ -164,6 +164,9 @@ final class NettyHttpRequest implements HttpRequest {
 
     @Override
     public HttpRequest appendPathPart(String encodedPathPart) {
+        // java-sdk follows whatwg spec for escaping, but RFC 2396 (followed by java.net.URI) does not permit pipe characters
+        encodedPathPart = encodedPathPart.replace("|", "%7c");
+
         boolean hasSlashLeft = uri.charAt(uri.length() - 1) == '/';
         boolean hasSlashRight = encodedPathPart.startsWith("/");
         if (hasSlashLeft) {

--- a/oraclecloud-httpclient-netty/src/test/groovy/io/micronaut/oraclecloud/client/OciCustomEndpointSpec.groovy
+++ b/oraclecloud-httpclient-netty/src/test/groovy/io/micronaut/oraclecloud/client/OciCustomEndpointSpec.groovy
@@ -47,6 +47,9 @@ class OciCustomEndpointSpec extends Specification {
         then:
         MockAuthenticationDetailsProvider == authenticationDetailsProvider.getClass()
         client.getEndpoint() == CUSTOM_ENDPOINT
+
+        cleanup:
+        context.close()
     }
 
     void "test custom region"() {
@@ -67,6 +70,9 @@ class OciCustomEndpointSpec extends Specification {
         then:
         MockAuthenticationDetailsProvider == authenticationDetailsProvider.getClass()
         client.getEndpoint() == "https://telemetry.us-ashburn-1.oraclecloud.com"
+
+        cleanup:
+        context.close()
     }
 
     void "test custom region factory"() {
@@ -86,6 +92,9 @@ class OciCustomEndpointSpec extends Specification {
         then:
         MockAuthenticationDetailsProvider == authenticationDetailsProvider.getClass()
         client.getEndpoint() == "https://telemetry.us-ashburn-1.oraclecloud.com"
+
+        cleanup:
+        context.close()
     }
 
     @Factory

--- a/oraclecloud-httpclient-netty/src/test/groovy/io/micronaut/oraclecloud/client/OciIdentitySpec.groovy
+++ b/oraclecloud-httpclient-netty/src/test/groovy/io/micronaut/oraclecloud/client/OciIdentitySpec.groovy
@@ -20,6 +20,7 @@ import static com.oracle.bmc.identity.requests.ListCompartmentsRequest.AccessLev
 @Requires(property = "vault.secrets.compartment.ocid")
 @Requires(bean = AuthenticationDetailsProvider)
 @MicronautTest
+@Property(name = "use.real.auth", value = "true")
 class OciIdentitySpec extends Specification {
 
     @Property(name = "vault.secrets.compartment.ocid")

--- a/oraclecloud-httpclient-netty/src/test/groovy/io/micronaut/oraclecloud/client/OciLoggingSpec.groovy
+++ b/oraclecloud-httpclient-netty/src/test/groovy/io/micronaut/oraclecloud/client/OciLoggingSpec.groovy
@@ -32,6 +32,7 @@ import java.time.temporal.ChronoUnit
 @Requires(bean = AuthenticationDetailsProvider)
 @MicronautTest
 @Stepwise
+@Property(name = "use.real.auth", value = "true")
 class OciLoggingSpec extends Specification {
 
     @Shared

--- a/oraclecloud-httpclient-netty/src/test/groovy/io/micronaut/oraclecloud/client/OciMonitoringSpec.groovy
+++ b/oraclecloud-httpclient-netty/src/test/groovy/io/micronaut/oraclecloud/client/OciMonitoringSpec.groovy
@@ -24,6 +24,7 @@ import java.time.temporal.ChronoUnit
 @Requires(property = "monitoring.compartment.ocid")
 @Requires(bean = AuthenticationDetailsProvider)
 @MicronautTest
+@Property(name = "use.real.auth", value = "true")
 class OciMonitoringSpec extends Specification {
 
     @Property(name = "monitoring.compartment.ocid")

--- a/oraclecloud-httpclient-netty/src/test/groovy/io/micronaut/oraclecloud/client/OciObjectStorageSpec.groovy
+++ b/oraclecloud-httpclient-netty/src/test/groovy/io/micronaut/oraclecloud/client/OciObjectStorageSpec.groovy
@@ -22,6 +22,7 @@ import spock.lang.Stepwise
 @Requires(bean = AuthenticationDetailsProvider)
 @MicronautTest
 @Stepwise
+@Property(name = "use.real.auth", value = "true")
 class OciObjectStorageSpec extends Specification {
 
     @Shared

--- a/oraclecloud-httpclient-netty/src/test/groovy/io/micronaut/oraclecloud/client/OciVaultSpec.groovy
+++ b/oraclecloud-httpclient-netty/src/test/groovy/io/micronaut/oraclecloud/client/OciVaultSpec.groovy
@@ -27,6 +27,7 @@ import java.time.temporal.ChronoUnit
 @Requires(bean = AuthenticationDetailsProvider)
 @MicronautTest
 @Stepwise
+@Property(name = "use.real.auth", value = "true")
 class OciVaultSpec extends Specification {
 
     @Shared

--- a/oraclecloud-httpclient-netty/src/test/java/io/micronaut/oraclecloud/httpclient/netty/MockAuthenticationDetailsProvider.java
+++ b/oraclecloud-httpclient-netty/src/test/java/io/micronaut/oraclecloud/httpclient/netty/MockAuthenticationDetailsProvider.java
@@ -13,9 +13,7 @@ import java.io.InputStream;
 @Singleton
 @Replaces(AuthenticationDetailsProvider.class)
 @Primary
-@Requires(missingProperty = "vault.secrets.compartment.ocid")
-@Requires(missingProperty = "vault.ocid")
-@Requires(missingProperty = "monitoring.compartment.ocid")
+@Requires(missingProperty = "use.real.auth")
 public class MockAuthenticationDetailsProvider implements AuthenticationDetailsProvider {
 
     @Override

--- a/oraclecloud-httpclient-netty/src/test/java/io/micronaut/oraclecloud/httpclient/netty/NettyHttpClientTest.java
+++ b/oraclecloud-httpclient-netty/src/test/java/io/micronaut/oraclecloud/httpclient/netty/NettyHttpClientTest.java
@@ -45,6 +45,8 @@ class NettyHttpClientTest {
             expected.put("foo", "bar");
             Assertions.assertEquals(expected, body);
         }
+        embeddedServer.close();
+        ctx.close();
     }
 
     @Test
@@ -65,6 +67,8 @@ class NettyHttpClientTest {
                     .streamBody().toCompletableFuture().get();
             Assertions.assertEquals("abcdef", new String(toByteArray(responseStream, 6), StandardCharsets.UTF_8));
         }
+        embeddedServer.close();
+        ctx.close();
     }
 
     private static byte[] toByteArray(InputStream stream, int n) throws IOException {
@@ -97,6 +101,8 @@ class NettyHttpClientTest {
                     .textBody().toCompletableFuture().get();
             Assertions.assertEquals("abcdef", response);
         }
+        embeddedServer.close();
+        ctx.close();
     }
 
     @Test
@@ -125,6 +131,8 @@ class NettyHttpClientTest {
                 });
             Assertions.assertNull(result.toCompletableFuture().get());
         }
+        embeddedServer.close();
+        ctx.close();
     }
 
     @Controller

--- a/oraclecloud-logging/src/test/groovy/io/micronaut/oraclecloud/logging/OracleCloudLoggingClientSpec.groovy
+++ b/oraclecloud-logging/src/test/groovy/io/micronaut/oraclecloud/logging/OracleCloudLoggingClientSpec.groovy
@@ -16,6 +16,9 @@ class OracleCloudLoggingClientSpec extends Specification {
 
         expect:
         !context.containsBean(OracleCloudLoggingClient)
+
+        cleanup:
+        context.close()
     }
 
 

--- a/oraclecloud-micrometer/src/test/groovy/io/micronaut/oraclecloud/monitoring/OracleCloudMeterRegistryFactorySpec.groovy
+++ b/oraclecloud-micrometer/src/test/groovy/io/micronaut/oraclecloud/monitoring/OracleCloudMeterRegistryFactorySpec.groovy
@@ -45,5 +45,8 @@ class OracleCloudMeterRegistryFactorySpec extends Specification {
 
         then:
         noExceptionThrown()
+
+        cleanup:
+        context.close()
     }
 }

--- a/oraclecloud-micrometer/src/test/groovy/io/micronaut/oraclecloud/monitoring/OracleCloudMeterRegistryFactoryTest.groovy
+++ b/oraclecloud-micrometer/src/test/groovy/io/micronaut/oraclecloud/monitoring/OracleCloudMeterRegistryFactoryTest.groovy
@@ -25,6 +25,9 @@ class OracleCloudMeterRegistryFactoryTest extends Specification {
         expect:
         !context.containsBean(OracleCloudMeterRegistry)
         !context.containsBean(OracleCloudRawMeterRegistry)
+
+        cleanup:
+        context.close()
     }
 
     def "test it not loads when disabled"() {
@@ -36,6 +39,9 @@ class OracleCloudMeterRegistryFactoryTest extends Specification {
         expect:
         !context.containsBean(OracleCloudMeterRegistry)
         !context.containsBean(OracleCloudRawMeterRegistry)
+
+        cleanup:
+        context.close()
     }
 
     def "test it loads by default"() {
@@ -47,6 +53,9 @@ class OracleCloudMeterRegistryFactoryTest extends Specification {
 
         expect:
         context.containsBean(OracleCloudMeterRegistry)
+
+        cleanup:
+        context.close()
     }
 
     def "test raw metrics meter registry"() {
@@ -60,6 +69,9 @@ class OracleCloudMeterRegistryFactoryTest extends Specification {
         expect:
         context.containsBean(OracleCloudRawMeterRegistry)
         !context.containsBean(OracleCloudMeterRegistry)
+
+        cleanup:
+        context.close()
     }
 
     def "test raw metrics meter registry loads when both are true"() {
@@ -74,6 +86,9 @@ class OracleCloudMeterRegistryFactoryTest extends Specification {
         expect:
         context.containsBean(OracleCloudRawMeterRegistry)
         !context.containsBean(OracleCloudMeterRegistry)
+
+        cleanup:
+        context.close()
     }
 
     def "test raw metrics meter registry loads when raw is true and aggregated is false"() {
@@ -91,5 +106,8 @@ class OracleCloudMeterRegistryFactoryTest extends Specification {
         context.containsBean(OracleCloudConfig)
         OracleCloudConfig oracleCloudConfig = context.getBean(OracleCloudConfig)
         oracleCloudConfig.enabled()
+
+        cleanup:
+        context.close()
     }
 }

--- a/test-suite-http-client/src/main/java/io/micronaut/oraclecloud/httpclient/NettyTest.java
+++ b/test-suite-http-client/src/main/java/io/micronaut/oraclecloud/httpclient/NettyTest.java
@@ -6,6 +6,7 @@ import com.oracle.bmc.auth.SimpleAuthenticationDetailsProvider;
 import com.oracle.bmc.common.ClientBuilderBase;
 import com.oracle.bmc.http.client.HttpClient;
 import com.oracle.bmc.http.client.HttpClientBuilder;
+import com.oracle.bmc.http.client.HttpRequest;
 import com.oracle.bmc.http.client.HttpResponse;
 import com.oracle.bmc.http.client.Method;
 import com.oracle.bmc.http.client.StandardClientProperties;
@@ -40,6 +41,7 @@ import java.io.FileNotFoundException;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.UncheckedIOException;
+import java.net.URLEncoder;
 import java.nio.charset.StandardCharsets;
 import java.security.cert.CertificateException;
 import java.time.Instant;
@@ -47,6 +49,7 @@ import java.util.ArrayList;
 import java.util.Date;
 import java.util.HashSet;
 import java.util.List;
+import java.util.Locale;
 import java.util.Set;
 import java.util.concurrent.ExecutionException;
 
@@ -655,6 +658,33 @@ public abstract class NettyTest {
                  .get()) {
             SessionTokenAuthenticationDetailsProvider.SessionToken resp = response.body(SessionTokenAuthenticationDetailsProvider.SessionToken.class).toCompletableFuture().get();
             Assertions.assertEquals("foo", resp.getToken());
+        }
+    }
+
+    @Test
+    public void pipeTest() throws Exception {
+        String testString = "foo|bar";
+        netty.handleOneRequest((ctx, request) -> {
+            Assertions.assertEquals(HttpMethod.GET, request.method());
+            Assertions.assertEquals("/" + URLEncoder.encode(testString, StandardCharsets.UTF_8).toLowerCase(Locale.ROOT), request.uri().toLowerCase(Locale.ROOT));
+
+            DefaultFullHttpResponse response = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.OK, Unpooled.wrappedBuffer("bar".getBytes(StandardCharsets.UTF_8)));
+            response.headers().add("Content-Type", "text/plain");
+            computeContentLength(response);
+            ctx.writeAndFlush(response);
+        });
+
+        try (HttpClient client = newBuilder()
+            .build()) {
+            HttpRequest request = client.createRequest(Method.GET)
+                .appendPathPart(testString);
+            Assertions.assertEquals("/" + testString, request.uri().getPath());
+            try (HttpResponse response = request
+                .execute().toCompletableFuture()
+                .get()) {
+                String s = response.textBody().toCompletableFuture().get();
+                Assertions.assertEquals("bar", s);
+            }
         }
     }
 

--- a/test-suite-java/src/test/java/example/HikariPoolAtpTest.java
+++ b/test-suite-java/src/test/java/example/HikariPoolAtpTest.java
@@ -49,5 +49,6 @@ class HikariPoolAtpTest {
         ResultSet resultSet = connection.createStatement().executeQuery("SELECT * FROM DUAL");
         resultSet.next();
         Assertions.assertEquals("X", resultSet.getString(1));
+        context.close();
     }
 }

--- a/test-suite-java/src/test/java/example/MicrometerTest.java
+++ b/test-suite-java/src/test/java/example/MicrometerTest.java
@@ -40,6 +40,7 @@ class MicrometerTest {
                 register(cloudMeterRegistry);
             counter.increment(5.0);
             Thread.sleep(2000);
+            context.close();
         });
     }
 

